File: javascript_code/graph/floodfill/MaxAreaOfIsland.js
export class MaxAreaOfIsland {

    constructor() {
        this.directions = [
            [-1, 0], [1, 0], [0, -1], [0, 1],
        ]
    }

    maxAreaOfIsland(grid) {
        if (grid == null) return 0

        this.rows = grid.length
        if (this.rows == 0) return 0
        this.cols = grid[0].length
        this.grid = grid

        // 1. 建图
        this.graph = this.constructGraph()

        // 2. 求解最大联通分量的顶点数
        this.visited = new Array(this.graph.length).fill(false)
        let res = 0
        for (let v = 0; v < this.graph.length; v++) {
            const row = Math.floor(v / this.cols)
            const col = v % this.cols
            if (!this.visited[v] && grid[row][col] == 1) {
                res = Math.max(this.dfs(v), res)
            }
        }
        return res
    }

    dfs(v) {
        this.visited[v] = true
        let res = 1
        for (const w of this.graph[v]) {
            if (!this.visited[w]) {
                res += this.dfs(w)
            }
        }
        return res
    }

    constructGraph() {
        const g = new Array(this.rows * this.cols).fill(0)
        for (let v = 0; v < g.length; v++) {
            g[v] = new Set()
        }

        for (let v = 0; v < g.length; v++) {
            const row = Math.floor(v / this.cols)
            const col = v % this.cols
            if (this.grid[row][col] == 1) {
                for (const dir of this.directions) {
                    const nextRow = row + dir[0]
                    const nextCol = col + dir[1]
                    if (this.inArea(nextRow, nextCol)
                        && this.grid[nextRow][nextCol] == 1) {
                        const w = nextRow * this.cols + nextCol
                        g[v].add(w)
                        g[w].add(v)
                    }
                }
            }
        }

        return g
    }

    inArea(row, col) {
        return row >= 0 && row < this.rows && col >= 0 && col < this.cols
    }

}

File: cpp_code/graph/floodfill/MaxAreaOfIsland.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

#include <vector>
#include <unordered_set>
#include <iostream>

using namespace std;

class MaxAreaOfIsland {
private:
    int rows;
    int cols;
    vector<vector<int>> grid;
    vector<vector<int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    vector<bool> visited;

    vector<unordered_set<int>> graph;

public:
    int maxAreaOfIsland(vector<vector<int>> grid) {
        rows = grid.size();
        if (rows == 0) return 0;

        cols = grid[0].size();
        this->grid = grid;

        // 1. 建图
        graph = constructGraph();

        // 2. 求解最大联通分量的顶点数
        this->visited.resize(graph.size());
        fill(this->visited.begin(), this->visited.end(), false);

        int res = 0;
        for (int v = 0; v < graph.size(); v++) {
            int row = v / cols;
            int col = v % cols;
            if (!visited[v] && grid[row][col] == 1) {
                res = max(dfs(v), res);
            }
        }
        return res;
    }

    int dfs(int v) {
        visited[v] = true;
        int res = 1;
        for (int w : graph[v]) {
            if (!visited[w]) {
                res += dfs(w);
            }
        }
        return res;
    }

    vector<unordered_set<int>> constructGraph() {
        vector<unordered_set<int>> g(rows * cols, unordered_set<int>());

        for (int v = 0; v < g.size(); v++) {
            int row = v / cols;
            int col = v % cols;
            if (grid[row][col] == 1) {
                for (vector<int> dir : directions) {
                    int nextRow = row + dir[0];
                    int nextCol = col + dir[1];
                    if (inArea(nextRow, nextCol)
                        && grid[nextRow][nextCol] == 1) {
                        int w = nextRow * cols + nextCol;
                        g[v].insert(w);
                        g[w].insert(v);
                    }
                }
            }
        }
        return g;
    }

    bool inArea(int row, int col) {
        return row >= 0 && row < rows && col >= 0 && col < cols;
    }

};

int main() {
    vector<vector<int>> grid = {
            {0,0,1,0,0,0,0,1,0,0,0,0,0},
            {0,0,1,0,0,0,0,1,1,1,0,0,0},
            {0,1,0,0,1,0,0,0,0,0,0,0,0},
            {0,1,0,0,1,1,0,0,1,0,1,0,0},
            {0,1,0,0,1,1,0,0,1,1,1,0,0},
            {0,0,1,0,0,0,0,0,0,0,1,0,0},
            {0,0,0,0,0,0,0,1,1,1,0,0,0},
            {0,0,0,0,0,0,0,1,1,0,0,0,0},
    };
    MaxAreaOfIsland maxAreaOfIsland;

    cout << to_string(maxAreaOfIsland.maxAreaOfIsland(grid));

    return 0;
}